Automattic/wp-calypso
# Store Products State: Merge search state into list state

In my work on the product search, I noticed that the products state is not very straightforward. I've tried to simplify it in this PR, by removing the separate search reducer in both `sites` & `ui`. There are details in [the products README](https://github.com/Automattic/wp-calypso/tree/update/store-product-search-state/client/extensions/woocommerce/state/sites/products) (and I've added a [README to the UI state too](https://github.com/Automattic/wp-calypso/tree/update/store-product-search-state/client/extensions/woocommerce/state/ui/products)).

These selectors have been removed:

```
areProductSearchResultsLoaded
areProductSearchResultsLoading
getTotalProductSearchResults
getProductSearchQuery
getProductListProducts
getProductSearchCurrentPage
getProductSearchResults
getProductSearchRequestedPage
```

These selectors have been added/renamed

```
  getProductsCurrentPage   (was getProductListCurrentPage)
+ getProductsCurrentSearch
  getProductsRequestedPage (was getProductListRequestedPage)
+ getProductsRequestedSearch
```

These are replaced with the more generic functions, which all expect a second argument with the request query. They default to a DEFAULT_QUERY set in [the new utils file](https://github.com/Automattic/wp-calypso/blob/update/store-product-search-state/client/extensions/woocommerce/state/sites/products/utils.js#L7-L11).

For example, some replacements:

- the `areLoading/ed` functions can be replaced with `areProductsLoading/ed( state, { search: term } )`. `getTotalProductSearchResults` can be replaced with `getTotalProducts( state, { search: term } )`.
- `getProductSearchQuery` has been removed from the sites state, and the search query is now stored in the UI state. Now, `getProductsCurrentSearch` will get the current search term.
- `getProductListProducts` & `getProductSearchResults` from `state/ui` are both replaced with `getProducts( state, params )` in `state/sites`.

These actions have been removed:

```
fetchProductSearchResults
clearProductSearch
```

You can now use `fetchProducts` directly with `{ search: term }` to get a search request. `clearProductSearch` has been removed entirely, since the property it reset doesn't exist anymore.

**To test**

- Run the tests: `npm run test-client client/extensions/woocommerce`
- Click through store to check that all products pages are loading correctly
- Make sure editing, saving, & deleting products still works 
- There should be no visual changes in this PR
